nemo/collections/nlp/data/language_modeling/megatron/dataset_utils.py

```diff
@@ -386,6 +388,118 @@ def create_masked_lm_predictions(
     return (output_tokens, masked_lm_positions, masked_lm_labels, token_boundary, masked_spans)
 
 
+def create_extreme_masked_lm_predictions(
+    tokens,
+    masked_lm_prob,
+    mask_id,
+    max_predictions_per_seq,
+    np_rng,
+    max_ngram_size=10,
+    min_ngram_size=2,
+    mean_ngram_size=5,
+    span_length_distribution="geometric",
+):
+    """Creates the predictions for the extreme span-masking UL2 objective.
+    Note: Tokens here are vocab ids and not text tokens."""
+    output_tokens = list(tokens)
+
+    masked_lm_positions = []
+    masked_lm_labels = []
+
+    num_to_predict = min(max_predictions_per_seq, max(1, int(round(len(tokens) * masked_lm_prob))))
+    # If the number of tokens to predict is less than the min ngram size, clam it to max predictions.
+    min_ngram_size = int(min(num_to_predict, min_ngram_size))
+
+    ngrams = np.arange(min_ngram_size, max_ngram_size + 1, dtype=np.int64)
+    if span_length_distribution == "uniform":
+        pvals = np.array([1.0 / (max_ngram_size - min_ngram_size + 1)] * (max_ngram_size - min_ngram_size + 1))
+
+    ngram_indexes = []
+    cand_indexes = [[i] for i in range(len(tokens))]
+    for idx in range(len(cand_indexes)):
+        ngram_index = []
+        for n in ngrams:
+            ngram_index.append(cand_indexes[idx : idx + n])
+        ngram_indexes.append(ngram_index)
```

Comment:
This is a little memory intensive. maybe considering using python generator.

Reply:
Agree. But it's not super big right now, for T5 we have ngrams from 1-10 and for UL2 it is maybe 32 - 64.